File: macos-native/GoalflowMac/Services/Clock.swift
import Foundation

/// Injected clock for deterministic timer logic (no decrement-integer drift).
protocol Clock: Sendable {
    func now() -> Date
}

struct SystemClock: Clock {
    func now() -> Date { Date() }
}

struct FixedClock: Clock {
    var fixed: Date
    func now() -> Date { fixed }
}

/// Mutable clock for tests.
final class ManualClock: Clock, @unchecked Sendable {
    private var _now: Date
    private let lock = NSLock()
    init(now: Date) { self._now = now }
    func now() -> Date { lock.lock(); defer { lock.unlock() }; return _now }
    func advance(by seconds: TimeInterval) {
        lock.lock(); _now = _now.addingTimeInterval(seconds); lock.unlock()
    }
    func set(_ date: Date) {
        lock.lock(); _now = date; lock.unlock()
    }
}

File: chrome-extension/src/services/Clock.ts
// Clock abstraction for deterministic timer tests — mirrors android-native/time/GoalflowTime.kt and macos-native/Services/Clock.swift

export interface Clock {
  now(): Date;
}

export class SystemClock implements Clock {
  now(): Date { return new Date(); }
}

export class FixedClock implements Clock {
  constructor(private readonly fixed: Date) {}
  now(): Date { return new Date(this.fixed.getTime()); }
}

export class ManualClock implements Clock {
  private ms: number;
  constructor(initial: Date) { this.ms = initial.getTime(); }
  now(): Date { return new Date(this.ms); }
  advance(ms: number): void { this.ms += ms; }
  advanceSeconds(s: number): void { this.ms += s * 1000; }
  set(date: Date): void { this.ms = date.getTime(); }
}
